### ISSUE TITLE
chap03_Linear_Regression_gradientTape

### ISSUE THREAD

**Issue by hih0124** [hih0124/lecture_deep_learning#27]
"""
tf.GradientTape + regression model
 -> 미분계수 자동 계산 -> model 최적화(최적의 기울기와 절편 update)
"""
import tensorflow as tf
tf.executing_eagerly() # 즉시 실행 test : ver2.x 사용중 

### 1. input/output 변수 정의 
inputs = tf.Variable([1.0, 2.0, 3.0]) # x변수 
outputs = tf.Variable([2.0, 4.0, 6.0]) # y변수

### 2. model : model 클래스
class Model(tf.keras.Model) : # tf.karas.Model 이용해서 자식클래스(부모클래스) 정의
    def __init__(self) : # 생성자 
        super().__init__() # 부모생성자 호출
        self.W = tf.Variable(tf.random.normal([1])) # 기울기(가중치)
        self.B = tf.Variable(tf.random.normal([1])) # 절편
    def call(self, inputs) : # 메서드 재정의
        return inputs * self.W + self.B # 회귀방정식(예측치)

-요런식으로 5번에서 모델 만들 것이야.
-model = Model()
-model.call(inputs) : call은 생략가능
-model(inputs)

### 2. 손실 함수
def loss(model, inputs, outputs):
  err = model(inputs) - outputs
  return tf.reduce_mean(tf.square(err)) # MSE

### 3. 미분계수(기울기) 계산  
def gradient(model, inputs, outputs) : 
    with tf.GradientTape() as tape:
        loss_value = loss(model, inputs, outputs) # 손실함수 호출  
        grad = tape.gradient(loss_value, [model.W, model.B])
        # 미분계수 -> 기울기와 절편 업데이트
    return grad # 업데이트 결과 반환

### 5. model 생성
model = Model() # 생성자
mse = loss(model, inputs, outputs)
print("mse= ", mse.numpy()) # mse=  55.1283

grad = gradient(model, inputs, outputs)
print("grad =", grad)
'''
grad = [<tf.Tensor: shape=(1,), dtype=float32,
 numpy=array([-3.6994476], dtype=float32)>, 
 <tf.Tensor: shape=(1,), dtype=float32, 
 numpy=array([-1.5804353], dtype=float32)>]
'''

### 6. model 최적화 객체
opt = tf.keras.optimizers.SGD(learning_rate = 0.01)
-학습 전 초기손실값과 w,b
print("초기 손실값 : {:.6f}".format(loss(model, inputs, outputs)))
print("w : {}, b : {}".format(model.W.numpy(), model.B.numpy()))
'''
초기 손실값 : 0.908750
w : [1.2683412], b : [2.206195]
'''

### 7. 반복학습
for step in range(300) :
    grad = gradient(model, inputs, outputs) # 기울기 계산
    -기울기 -> 최적화 객체 반영
    opt.apply_gradients(zip(grad, [model.W, model.B]))
    
    if (step+1) % 20 ==0 :
        print("step = {}, loss = {:.6f}".format((step+1), loss(model, inputs, outputs)))
    
'''
step = 296, loss = 0.023121
step = 297, loss = 0.023010
step = 298, loss = 0.022899
step = 299, loss = 0.022789
step = 300, loss = 0.022680
'''

- model 최적화
print("최종 손실값 : {:.6f}".format(loss(model, inputs, outputs)))
print("w : {}, b : {}".format(model.W.numpy(), model.B.numpy()))
'''
최종 손실값 : 0.151871
w : [1.5473804], b : [1.0289104]
'''   

- model test
y_pred = model.call(2.5) # x = 2.5  
print("y_pred", y_pred.numpy()) # y_pred [4.977716]
